pyro/infer/predictive.py

```diff
@@ -219,3 +231,82 @@ def get_vectorized_trace(self, *args, **kwargs):
                                             parallel=self.parallel, model_args=args, model_kwargs=kwargs)
         return _predictive(self.model, posterior_samples, self.num_samples,
                            return_trace=True, model_args=args, model_kwargs=kwargs)
+
+
+def log_likelihood(
+    model: Callable,
+    posterior_samples: Dict[str, torch.Tensor] = None,
+    guide: Callable = None,
+    num_samples: int = None,
+    parallel: bool = True,
+) -> Callable:
+    """EXPERIMENTAL utility to calculate the log likelihood of the data for a
+    model with observed sites, given posterior samples (or a guide that can
+    generate them).
+
+    :param model: Model function containing pyro primitives. Must have observed
+        sites conditioned on data.
+    :type model: Callable
+    :param posterior_samples: Dictionary of posterior samples, defaults to None
+    :type posterior_samples: Dict[str, torch.Tensor], optional
+    :param guide: Guide function, defaults to None
+    :type guide: Callable, optional
+    :param num_samples: Number of posterior samples to evaluate, defaults to None
+    :type num_samples: int, optional
+    :param parallel: Whether to use vectorization, defaults to True
+    :type parallel: bool, optional
+    :rtype: Callable
+
+    The API parallels that of the :class:`Predictive` interface (assume `model`
+    and `guide` have already been defined elsewhere):
+
+    ```
+    x = torch.randn(10)
+    y = x + 0.05 * torch.randn(10)
+    cond_model = pyro.condition(model, data={"y": y})
+    ll_vals = log_likelihood(cond_model, guide=guide, num_samples)(x)
+    ```
+    """
+
+    _validate_predictive_inputs(posterior_samples, num_samples, guide, return_sites=None)
+
+    # define the function to return the likelihood values
+    def log_like_helper(*args, **kwargs) -> torch.Tensor:
+        # trace model once to get observed nodes
+        trace = pyro.poutine.trace(model).get_trace(*args, **kwargs)
```

Comment:
You can avoid tracing the model to get the observed node by using
```
collect_fn=lambda msg: msg["fn"].log_prob(msg["value"]) if msg["is_observed"] else None
```

Reply:
I'm not sure that I can get rid of the trace in the parallel branch because it seems that `get_vectorized_trace` is returning `site["is_observed"] == True` even for sites that are not actually observed. I replaced
```python
for obs_name in observed.keys():
    site = trace.nodes[obs_name]
    log_like[obs_name] = site["fn"].log_prob(site["value"])
```
with
```python
for obs_name, site in trace.nodes.items():
    if site["type"] == "sample" and site["is_observed"]:
        log_like[obs_name] = site["fn"].log_prob(site["value"])
```
and my tests start failing because there are more observed nodes being returned than expected. It seems the problem is the following line in `_predictive`:
```python
if return_trace:
    trace = poutine.trace(poutine.condition(vectorize(model), reshaped_samples))\
        .get_trace(*model_args, **model_kwargs)
    return trace
```
where once the model is conditioned on the samples, then the trace thinks that they are now observed. Perhaps I can just extract the parts of `_predictive` that are necessary for getting the trace and modify this part?

Reply:
Also, in the serial branch, I think we need to do a trace before passing to `_predictive_sequential` so that we can get `site_shapes` to pass in as an argument. Maybe it might be better to unpack the functionality that we need from `_predictive_sequential` and rewrite this part too?

Reply:
It looks like each call to `get_vectorized_trace` already traces the model three times (in `_predictive`, once to guess plate nesting, once to prune subsamples, third time to actually return the trace); six times total if you provide a guide instead of samples. Maybe adding one more trace isn't so bad haha.

Reply:
re: my first comment - it looks like this behavior would also occur in `_predictive_sequential` as well; there are some lines like
```python
for i in range(num_samples):
    trace = poutine.trace(poutine.condition(model, samples[i])).get_trace(*model_args, **model_kwargs)
    if return_trace:
        collected.append(trace)
    else:
        collected.append({site: collect_fn(trace.nodes[site]) for site in return_site_shapes})
```
the only thing that prevents this from returning extra observed sites at the moment is that `return_site_shapes` only has keys from observed nodes (which are gathered from the trace in `log_likelihood_helper`) so the extras get filtered out in the last line

Reply:
Whoa, it is tricky... Thanks for explaining, @ecotner! Let's keep this as-is. :)

Sorry for making confusion!

Reply:
no worries; no confusion here. it is just a tough problem to solve haha. I think a genuine solution would involve rewriting large parts of `_predictive` and `_predictive_sequential` though, and I worry about accidentally breaking any downstream functionality that relies on this (like `Predictive` or utilities in `arviz`).

Reply:
oh, you know what would be a real simple fix? we could just have the user pass in the `observed` dictionary themselves as an optional argument. if it is not present, then we do the extra trace to generate it.

Reply:
I think it is better to keep the previous interface. There are several ways to get around this issue but it seems unnecessary here (in NumPyro, we also support computing log likelihood for 1 sample so avoiding tracing many times is needed - there, we use [substitute](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer/util.py#L615) rather than `condition`, so `is_observed` is not overwritten after substituted).

If you want to avoid tracing here, then you can use [infer_config_messenger](https://github.com/pyro-ppl/pyro/blob/dev/pyro/poutine/infer_config_messenger.py) to add `_is_observed` keyword to `msg["infer"]`. Then, in `collect_fn`, you only compute log prob if `msg["infer"].get("_is_observed") is True`. Similarly, in `parallel` branch, after getting the vectorized trace, you only compute log prob at sites that satisfy `msg["infer"].get("_is_observed") is True`. Does this make sense at all? I think it will work but not test it yet. Please let me know if you want me to try some code here. :)

Reply:
yeah that makes sense. is `substitute` basically the same thing as [`do`](http://docs.pyro.ai/en/stable/poutine.html#pyro.poutine.handlers.do)? could an alternative be to possibly replace the `condition` calls in `_predictive` and `_predictive_sequential` with `do`? yeah feel free to experiment if you want!

Reply:
Yes, it works like `do` but does not introduce a fresh sample site. I think it is a bit expensive here if we use `do`.

Reply:
@ecotner I think it is better to use the previous API and leave comments for optimization in future PRs because they are not important if we use log_likelihood for a large number of samples (says 1000).